cli/graph_util.rs

```diff
@@ -320,14 +320,14 @@ struct GraphData {
 /// Holds the `ModuleGraph` and what parts of it are type checked.
 #[derive(Clone)]
 pub struct ModuleGraphContainer {
-  update_semaphore: Arc<Semaphore>,
+  update_lock: Arc<Mutex<()>>,
```

Comment:
I feel like using a semaphore is more applicable here because the data is not behind the mutex and using a Mutex here makes it seem like the data should go into it instead of being readonly sychronously (which I don’t believe Mutex supports quite like this). Internally IIRC, tokio’s Mutex is just `Semaphore::new(1)` (I’m just on my phone). The unwrap is not a big deal because that would only panic if the semaphore is closed.

So overall, I don’t think we should use a mutex since it doesn’t convey intent as well as a semaphore in my opinion.

Reply:
I don't know... on my first look at the new container `Semaphore` made it seem more complicated than it turned out to be and a mutex would have been more digestible. Semaphore=1 has capability in terms disregarding ownership and more which you have to look around for, my understanding is to practically prefer mutexes if you're not using those. I think it has less to do with being technically contained in the mutex since most languages don't have that. This is 6 lines of encapsulated logic either way, I just think it could be less of a stop-and-look.

Reply:
A semaphore with 1 permit is a pattern I’ve come across a lot. For me, mutex is more confusing. Here’s the example with mutex using a semaphore internally: https://github.com/tokio-rs/tokio/blob/ff2f286c120680f03182082a3fb97d69cd71d27d/tokio/src/sync/mutex.rs#L362

Reply:
I opened #18014. It should be clear now and it extracts out the other usages of `Semaphore::new(1)`